### PR TITLE
SIMSBIOHUB-602: Include Animals in the Survey page map

### PR DESCRIPTION
**Pull request by mauberti-bc**
# Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-602 (for the map popups)

# Description of changes

- Added ```ajv``` dependency to fix a dependency issue (should it be in overrides?)
- Added Animals to the Survey page data

Related PR that fixes the detailed critter endpoint SQL: https://github.com/bcgov/critterbase-api/pull/60 (merged)

# Testing Notes

On Survey page, the observation, animal, and telemetry data should appear in the map and the table below. All items on the map should have popups. The sample site/study area additionally have a tooltip.

### CODE REVIEW

File: app/src/utils/Utils.tsx [hunk 1]
```diff
@@ -478,3 +478,40 @@ export const getRandomHexColor = (seed: number, min = 100, max = 170): string =>
  * @return {*}  {value is T}
  */
 export const isDefined = <T,>(value: T | undefined | null): value is T => value !== undefined && value !== null;
+
+/**
+ * Creates a GeoJSON point from a set of coordinates and associated properties.
+ *
+ * @param latitude - The latitude coordinate.
```
Comment (MacDeluca) [inline app/src/utils/Utils.tsx]: The jsdoc syntax looks slightly off @param {number} latitude - The latitude coordinate.

File: app/src/utils/Utils.tsx [hunk 2]
```diff
@@ -478,3 +478,40 @@ export const getRandomHexColor = (seed: number, min = 100, max = 170): string =>
...
+ * @param longitude - The longitude coordinate.
+ * @returns A GeoJSON point
+ */
+export const createGeoJSONPoint = (latitude: number, longitude: number): Point => {
```
Comment (MacDeluca) [inline app/src/utils/Utils.tsx]: Maybe these helper functions could be moved to its own utils file. utils/spatial or utils/geoJson

File: app/src/hooks/useContext.tsx [hunk 3]
```diff
@@ -147,6 +148,23 @@ export const useObservationsTableContext = (): IObservationsTableContext => {
   return context;
 };
 
+/**
+ * Returns an instance of `IObservationsDataContext` from `ObservationsDataContext`.
```
Comment (MacDeluca) [inline app/src/hooks/useContext.tsx]: TelemetryDataContext

File: app/src/hooks/api/useSurveyApi.ts [hunk 4]
```diff
@@ -399,6 +400,22 @@ const useSurveyApi = (axios: AxiosInstance) => {
     return data;
   };
 
+  /**
+   * Retrieve a list of critters associated with the given survey with details taken from critterbase, including
+   * additiona information such as captures and mortality
```
Comment (MacDeluca) [inline app/src/hooks/api/useSurveyApi.ts]: additional**

File: app/src/features/surveys/view/components/spatial-data/components/telemetry/table/SurveyDataTelemetryTable.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+const SurveyDataTelemetryTable = (props: ISurveyDataTelemetryTableProps) => {
+  const surveyContext = useContext(SurveyContext);
+
+  /**
+   * Memoized calculation of table rows based on critter deployments data.
+   * Formats dates and combines necessary fields for display.
+   */
+  const tableRows: ITelemetryData[] = useMemo(() => {
+    return surveyContext.critterDeployments.map((item) => {
+      console.log(item);
```
Comment (MacDeluca) [inline app/src/features/surveys/view/components/spatial-data/components/telemetry/table/SurveyDataTelemetryTable.tsx]: 👀

File: app/src/constants/spatial.ts [hunk 6]
```diff
@@ -21,9 +21,10 @@ export const ALL_OF_BC_BOUNDARY: Feature = {
 };
 
 export const SURVEY_MAP_LAYER_COLOURS = {
```
Comment (MacDeluca) [inline app/src/constants/spatial.ts]: This can be moved to constants/colours

File: app/src/features/surveys/view/SurveyPage.tsx [hunk 7]
```diff
@@ -36,13 +35,10 @@ const SurveyPage: React.FC = () => {
       <Container maxWidth="xl" sx={{ py: 3 }}>
         <Stack gap={3}>
           <TaxonomyContextProvider>
-            <SurveySpatialData />
+            {/* <SurveySpatialData /> */}
```
Comment (MacDeluca) [inline app/src/features/surveys/view/SurveyPage.tsx]: 👀

File: api/src/services/critterbase-service.ts [hunk 8]
```diff
@@ -384,6 +389,11 @@ export class CritterbaseService {
     return response.data;
   }
 
+  async getMultipleCrittersByIdsDetailed(critter_ids: string[]): Promise<ICritterDetailed[]> {
```
Comment (MacDeluca) [inline api/src/services/critterbase-service.ts:569]: jsdoc

File: app/src/features/surveys/view/components/spatial-data/components/animal/SurveyDataAnimal.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  const capturePoints: ISurveyMapPoint[] = useMemo(() => {
+    const points: ISurveyMapPoint[] = [];
+
+    // Iterate over animals array to extract capture points
+    for (const animal of animals ?? []) {
+      // Iterate over captures array within each animal
+      for (const capture of animal.captures ?? []) {
+        if (capture.capture_location?.latitude && capture.capture_location?.longitude) {
+          const point: ISurveyMapPoint = {
+            feature: {
```
Comment (MacDeluca) [inline app/src/features/surveys/view/components/spatial-data/components/animal/SurveyDataAnimal.tsx]: createGeoJsonFeature?

File: app/src/features/surveys/view/components/spatial-data/components/telemetry/SurveyDataTelemetry.tsx [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+                  coordinates: [telemetry.longitude, telemetry.latitude] as Position
+                }
+              },
+              key: `telemetry-${telemetry.telemetry_manual_id}`,
+              onLoadMetadata: () => formatTelemetryMetadata(telemetry, deployment, critter)
+            };
+          }) ?? []
+      );
+    },
+    []
```
Comment (MacDeluca) [inline app/src/features/surveys/view/components/spatial-data/components/telemetry/SurveyDataTelemetry.tsx]: Dependencies?

**Comment by MacDeluca** [comment]
Bug fix:
1. Clicking on capture point on the map correctly loads the data first time, second click renders skeletons, third click works.

Some ideas (feel free to include whichever ones):
1. We could disable the tool tip when we select a map element (to prevent the tooltip rendering underneath). Hover to render the tooltip -> click to see details (only one rendering at a time).
image
2. Tooltips missing for observations and animals / captures

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+      // Exit early if there are no critters
+      if (critters.length < 1) {
+        return res.status(200).json({ captures: [], mortalities: [] });
+      }
+
+      const captures: ICaptureGeometry[] = [];
+
+      // Map over all captures to extract geometry as points
+      for (const critter of critters) {
```
Comment (MacDeluca) [inline api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts]: Should this logic be moved to its own service method?

Comment (NickPhura) [reply api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts]: It looks like this endpoint isn't ever called. So might be able to remove it entirely.

**Comment by NickPhura** [comment]
> Bug fix:
> 
>     1. Clicking on capture point on the map correctly loads the data first time, second click renders skeletons, third click works.
> 

@MacQSL Pushed up a fix for this

Also added tooltips to the observation/animal/telemetry points.

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts (hunk 11 above)
Comment (NickPhura) [reply api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts]: In the meantime, I've put a TODO on the endpoint to flag that its unused, and can maybe be removed.

**Review by MacDeluca — DISMISSED** [review]
LGTM

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts (hunk 11 above)
Comment (mauberti-bc) [reply api/src/paths/project/{projectId}/survey/{surveyId}/critters/spatial.ts]: I created `/critters/spatial` in SIMS before realizing it should be Critterbase with `/critters/spatial`, so yes, no problem removing it

**Review by NickPhura — APPROVED** [review]
🌵

**Pull request merged**